addon/openssl-crypto.cc

```diff
@@ -0,0 +1,457 @@
+    const EVP_MD* hash = Hash::Get();
+    ASSERT(hash);
+
+    EVP_MD_CTX* digest_ctxp = S(EVP_MD_CTX_new)();
+    auto cleanup_ctx = Cleanup([&]() { S(EVP_MD_CTX_free)(digest_ctxp); });
+
+    if (!S(EVP_DigestInit_ex)(digest_ctxp, hash, nullptr)) {
+        return set_status_from_openssl(status, "error in EVP_DigestInit_ex");
+    }
+
+    if (!S(EVP_DigestUpdate)(digest_ctxp, static_cast<unsigned char*>(in->data), in->len)) {
+        return set_status_from_openssl(status, "error in EVP_DigestUpdate");
+    }
+
+    if (!S(EVP_DigestFinal_ex)(digest_ctxp, static_cast<unsigned char*>(out->data), nullptr)) {
+        return set_status_from_openssl(status, "error in EVP_DigestFinal_ex");
+    }
+
+    return true;
+}
+
+template <typename Hash>
+bool sign_rsa(void* unused_ctx,
+              mongocrypt_binary_t* key,
+              mongocrypt_binary_t* in,
+              mongocrypt_binary_t* out,
+              mongocrypt_status_t* status) {
+    const EVP_MD* hash = Hash::Get();
+
+    ASSERT(hash);
+    ASSERT(key);
+    ASSERT(in);
+    ASSERT(out);
+    ASSERT(status);
+
+    EVP_PKEY* pkey = nullptr;
+    size_t signature_out_len = 256;
+
+    EVP_MD_CTX* ctx = S(EVP_MD_CTX_new)();
+    auto cleanup_ctx = Cleanup([&]() { S(EVP_MD_CTX_free)(ctx); });
+    ASSERT(key->len <= LONG_MAX);
+    const unsigned char* key_data = static_cast<unsigned char*>(key->data);
+    pkey = S(d2i_PrivateKey)(EVP_PKEY_RSA, nullptr, &key_data, static_cast<long>(key->len));
+    auto cleanup_pkey = Cleanup([&]() { S(EVP_PKEY_free)(pkey); });
+    if (!pkey) {
+        return set_status_from_openssl(status, "error in d2i_PrivateKey");
+    }
+
+    if (!S(EVP_DigestSignInit)(ctx, nullptr, hash, nullptr /* engine */, pkey)) {
+        return set_status_from_openssl(status, "error in EVP_DigestSignInit");
+    }
+
+    if (!S(EVP_DigestSignUpdate)(ctx, static_cast<unsigned char*>(in->data), in->len)) {
+        return set_status_from_openssl(status, "error in EVP_DigestSignUpdate");
+    }
+
+    if (!S(EVP_DigestSignFinal)(ctx, static_cast<unsigned char*>(out->data), &signature_out_len)) {
+        return set_status_from_openssl(status, "error in EVP_DigestSignFinal");
+    }
+
+    return true;
+}
+
+void* opensslsym(const char* name) {
+    static struct OwnProcessDylib {
+        bool initialized = false;
+#ifdef _WIN32
+        HMODULE lib;
+
+        OwnProcessDylib() {
+            lib = GetModuleHandle(nullptr);
+        }
+
+        void* sym(const char* name) {
+            return reinterpret_cast<void*>(reinterpret_cast<uintptr_t>(GetProcAddress(lib, name)));
+        }
+#else
+        void* lib = nullptr;
+
+        OwnProcessDylib() {
+            lib = dlopen(nullptr, RTLD_NOW);
+        }
+
+        ~OwnProcessDylib() {
+            dlclose(lib);
+        }
+
+        void* sym(const char* name) {
+            return reinterpret_cast<void*>(dlsym(lib, name));
+        }
+#endif
+    } dl;
+    if (!dl.lib) {
```

Comment:
It's not a particularly odd thing to do imo – it's a "convenient" way of saying that this struct will be used in this place and in this place only, and that gives a bit of flexibility – in this case, for example, it breaks [the 0/3/5 rule](https://en.cppreference.com/w/cpp/language/rule_of_three), which would be bad if there was any risk of it being used in a situation in which that rule becomes relevant.

So, while it's easy to move this out of the function, and I'm happy to do so if you think that that's more readable, it would add a bit more boilerplate code here as well (deleting the move/copy constructors and operators).

Reply:
Cool, thanks.  To be honest I think what threw me off more than the inline struct definition was the combination of the inline definition _with_ the unfamiliar class construction syntax.  `<type> name;` calling the default constructor always throws me off because I'm used to `new <type>()` instead 🙂 

I do use inline type definitions in TS for the same reason, so fine with me if we leave this here.  Just to clarify - I was just thinking of leaving the struct in opensslsym:
```c++
void* opensslsym(const char* name) {
    struct OwnProcessDylib {
        ...
    };
    static OwnProcessDylib dl;
    if (!dl.lib) {
        throw new std::runtime_error("Could not open process handle");
    }

    return dl.sym(name);
}
```

Reply:
Ah sure, that doesn't make much of a difference – done in 414f95e216316c0d147659605f93fb21688a79b5 🙂